sparse checkout feature

officialasishkumar opened the pull request:
<!-- Thank you for contributing to KCL!

Note: 

1. With pull requests:

    - Open your pull request against "main"
    - Your pull request should have no more than two commits, if not you should squash them.
    - It should pass all tests in the available continuous integration systems such as GitHub Actions.
    - You should add/modify tests to cover your proposed code changes.
    - If your pull request contains a new feature, please document it on the README.

2. Please create an issue first to describe the problem.

    We recommend that link the issue with the PR in the following question.
    For more info, check https://kcl-lang.io/docs/community/contribute/
-->

#### 1. Does this PR affect any open issues?(Y/N) and add issue references (e.g. "fix #123", "re #123".):

- [ ] N
- [ ] Y 

<!-- You can add issue references here. 
    e.g. 
    fix #123, re #123, 
    fix https://github.com/XXX/issues/44
-->

#### 2. What is the scope of this PR (e.g. component or file name):

<!-- You can add the scope of this change here. 
    e.g. 
    /src/server/core.rs,
    kcl-lang/kcl/src/parser 
-->

#### 3. Provide a description of the PR(e.g. more details, effects, motivations or doc link):

<!-- You can choose a brief description here -->
- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Other

<!-- You can add more details here.
    e.g. 
    Call method "XXXX" to ..... in order to ....,
    More details: https://XXXX.com/doc......
-->

#### 4. Are there any breaking changes?(Y/N) and describe the breaking changes(e.g. more details, motivations or doc link):

- [ ] N
- [ ] Y 

<!-- You can add more details here.
    e.g. 
    Calling method "XXXX" will cause the "XXXX", "XXXX" modules to be affected.
    More details: https://XXXX.com/doc......
-->

#### 5. Are there test cases for these changes?(Y/N) select and add more details, references or doc links:

<!-- You can choose a brief description here -->
- [ ] Unit test
- [ ] Integration test
- [ ] Benchmark (add benchmark stats below)
- [ ] Manual test (add detailed scripts or steps below)
- [ ] Other

<!-- You can add more details here.
e.g. 
The test case in XXXX is used to .....
test cases in /src/tests/XXXXX
test cases https://github.com/XXX/pull/44
benchmark stats: time XXX ms
-->

officialasishkumar commented:
PTAL @zong-zhe @Peefy

officialasishkumar commented:
By mistake i defined the package in https://github.com/kcl-lang/kcl-go/blob/1460c5ca64b0d9ef3c10d7a7b1dcc7f98991bd87/pkg/kcl/opt.go#L15 and it worked locally. I didn't realised that and pushed it. I was wondering do i have to make changes to that repo too?

zong-zhe commented:
> By mistake i defined the package in https://github.com/kcl-lang/kcl-go/blob/1460c5ca64b0d9ef3c10d7a7b1dcc7f98991bd87/pkg/kcl/opt.go#L15 and it worked locally. I didn't realised that and pushed it. I was wondering do i have to make changes to that repo too?
> 
> There is another option to pass package here https://github.com/kcl-lang/kcl-go/blob/1460c5ca64b0d9ef3c10d7a7b1dcc7f98991bd87/pkg/spec/gpyrpc/gpyrpc.pb.go#L1497
> 
> lmk which sounds good
> 
> cc: @Peefy @zong-zhe

Hi @officialasishkumar 😄 

At present, define the `package` field in the kpm project because the current definition of KCL-GO is the go-sdk of KCL language; from the language side, the compiler can not understand the concept of `package`; all package-related content should be managed by kpm. Although they all work well, having a clear division of responsibility for each project is more helpful.

officialasishkumar commented:
Hi @zong-zhe , 

Thanks for the response. I have changed the code accordingly. Please have a look : )

zong-zhe commented on pkg/utils/utils.go:
```diff
@@ -599,3 +599,35 @@ func AbsTarPath(tarPath string) (string, error) {
+			return err
+		}
+
+		if info.IsDir() && strings.EqualFold(info.Name(), targetFolder) {
```
Hi @officialasishkumar 😄 

This function needs to look for the package name, not the name of the file directory, here you need to compare the two points, 1. Check whether `kcl.mod` exists in the file path. 2. Whether the package name in `kcl.mod` is the package name you are looking for.

```
if <there is a kcl.mod under path> && <the package name in kcl.mod matched> {
            return <the kcl.mod dir>
}
```

officialasishkumar replied:
> Hi @officialasishkumar 😄 
> 
> This function needs to look for the package name, not the name of the file directory, here you need to compare the two points, 1. Check whether `kcl.mod` exists in the file path. 2. Whether the package name in `kcl.mod` is the package name you are looking for.
> 
> ```
> if <there is a kcl.mod under path> && <the package name in kcl.mod matched> {
>             return <the kcl.mod dir>
> }
> ```
> 

wdym by package name

officialasishkumar replied:
I have updated the file accordingly. Please have a look @Peefy @zong-zhe!

officialasishkumar commented:
Thanks for the review. Will look into it!

> I am not sure if this PR can meet all the requirements of LFX,

Can you please clarify all the requirements so that I can start working on it?

officialasishkumar commented on pkg/utils/utils.go:
```diff
@@ -599,3 +600,48 @@ func AbsTarPath(tarPath string) (string, error) {
+	}
+
+	var modFile ModFile
+	_, err := toml.DecodeFile(kclModPath, &modFile)
```
Where is it? Can you share the link of the code or the docs?

officialasishkumar commented:
@peefy

The implementation of my design docs was in https://github.com/kcl-lang/kpm/pull/387

officialasishkumar commented on pkg/utils/utils.go:
```diff
@@ -599,3 +600,48 @@ func AbsTarPath(tarPath string) (string, error) {
+			return err
+		}
+		if info.IsDir() {
+			kclModPath := filepath.Join(path, "kcl.mod")
```
done!

officialasishkumar commented:
Sure thing, I have updated the title. 

> kcl mod add <sub_package> --git foo

Since the idea was changed that is to download the entire repo, I asked in slack whether to use package command in `add` or `run`. I found `run` to be more suitable since the repo is download entirely and the user can change it whenever they want while running their program. So, I implemented it the same way.

officialasishkumar commented:
Okay, So which idea should I write down in the desing doc? 

`kcl mod run --package test` 

or 

`kcl mod add --git <git-url> --commit <commit-hash> --package test`

I will update the code accordingly.

officialasishkumar commented:
No @Peefy
